runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/WorkerStatusClient.java

```diff
@@ -0,0 +1,165 @@
+    return new WorkerStatusClient(workerId, requestObserver);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness. A random id will
+   * be used to specify the request_id field.
+   *
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus() {
+    WorkerStatusRequest request =
+        WorkerStatusRequest.newBuilder().setId(idGenerator.getId()).build();
+    return getWorkerStatus(request);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness with request.
+   *
+   * @param request WorkerStatusRequest to be sent to SDK Harness.
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  CompletableFuture<WorkerStatusResponse> getWorkerStatus(WorkerStatusRequest request) {
+    CompletableFuture<WorkerStatusResponse> future = new CompletableFuture<>();
+    this.responseQueue.put(request.getId(), future);
+    this.requestReceiver.onNext(request);
+    return future;
+  }
+
+  /**
+   * Set up a deregister call back function for cleaning up connected client cache.
+   *
+   * @param deregisterCallback Consumer that takes worker id as param for deregister itself from
+   *     connected client cache when close is called.
+   */
+  public void setDeregisterCallback(Consumer<String> deregisterCallback) {
+    this.deregisterCallback = deregisterCallback;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (isClosed.getAndSet(true)) {
+      return;
+    }
+    for (CompletableFuture<WorkerStatusResponse> pendingResponse : responseQueue.values()) {
+      pendingResponse.completeExceptionally(
+          new RuntimeException("Fn Status Api client shut down while waiting for the request"));
+    }
+    responseQueue.clear();
+    requestReceiver.onCompleted();
+    if (deregisterCallback != null) {
+      deregisterCallback.accept(workerId);
+    }
+  }
+
+  /** Check if the client connection has already been closed. */
+  public boolean isClosed() {
+    return isClosed.get();
+  }
+
+  /** Get the worker id for the client's corresponding SDK Harness. */
+  public String getWorkerId() {
+    return this.workerId;
+  }
+
+  /** Get the response observer of this client for retrieving inbound worker status responses. */
+  public StreamObserver<WorkerStatusResponse> getResponseObserver() {
+    return new ResponseStreamObserver();
+  }
+
+  /**
+   * ResponseObserver for handling status responses. Each request will be cached with it's
+   * request_id. Upon receiving response from SDK Harness with this StreamObserver, the future
+   * mapped to same request_id will be finished accordingly.
+   */
+  private class ResponseStreamObserver implements StreamObserver<WorkerStatusResponse> {
+
+    @Override
+    public void onNext(WorkerStatusResponse response) {
+      if (!responseQueue.containsKey(response.getId())) {
```

Comment:
Handle concurrency race between `containsKey` and `remove` for the same id.
```
CompletableFuture<WorkerStatusResponse> future = responseQueue.remove(response.getId());
if (response != null) {
  future.complete(response);
} else {
  LOG.warn(String.format("Received response for status with unknown id %s and status %s", ...));
}
```

Reply:
applied change.